### ISSUE TITLE
[FEA] Ability to read ORC date and timestamp columns as GDF_DATE32 and GDF_TIMESTAMP

### ISSUE THREAD

**Issue by jlowe** [rapidsai/cudf#2384]
Currently the ORC loader forces a promotion of date and timestamp columns to GDF_DATE64 in the following code snippet:
```
    case orc::TIMESTAMP:
      // There isn't a GDF_TIMESTAMP -> np.dtype mapping so use np.datetime64
      return std::make_pair(GDF_DATE64, gdf_dtype_extra_info{TIME_UNIT_ms});
    case orc::DATE:
      // There isn't a GDF_DATE32 -> np.dtype mapping so use np.datetime64
      return std::make_pair(GDF_DATE64, gdf_dtype_extra_info{TIME_UNIT_ms});
```

For the Spark use-case, these values should be passed through as GDF_DATE32 and GDF_TIMESTAMP, respectively.

**Describe the solution you'd like**
Adding a flag in `orc::reader_options` to control whether these ORC types are mapped to their natural cudf types or to numpy-compatible types.

**Describe alternatives you've considered**
The ORC reader could always match these to the natural cudf column type, and the Python layer could run a type conversion kernel to make it numpy compatible when necessary.  However this incurs an extra kernel call and additional intermediate column storage.  A similar situation would exist for Spark if the conversion is required after the load.

**Comment by OlivierNV** [comment]
Sounds reasonable. I wonder if we could detect spark vs pandas (IIRC, both use a custom metadata string, so we could also have a 'smart' defaults that defaults to numpy-compatible dtypes if it detects the pandas metadata string).
@j-ieong  what do you think ? (low-level kernel already supports outputting either date32/date64, and afaik there are no difference between GDF_TIMESTAMP and GDF_DATE64 other than it's a different gdf enum from the reader's point of view.

**Comment by jlowe** [comment]
>  I wonder if we could detect spark vs pandas

I would avoid trying to auto-detect.  I could see workflows where datasets created by Spark are loaded into Pandas or vice-versa, and this behavior is all about the reader rather than the writer.  I agree this auto-detect logic probably "just works" most of the time, but it would be surprising for the cases where it doesn't.

I'd rather see an explicit setting.  That hopefully makes the caller think about how that setting would affect them so they set it properly, and the behavior would be consistent across all datasets.  Then data loading would always work as expected rather than work until it doesn't.

**Comment by OlivierNV** [comment]
@jlowe : I'm assuming that this is for the C++ API only (reader_options flag), or do you also need this flag available on the python side ?

**Comment by jlowe** [comment]
This would only be for the C++ API.  I would assume, at least for now, that everything on the Python side wants data that plays nice with numpy.

**Comment by mjsamoht** [comment]
Fixed with #2437.
